### PR TITLE
7903344: refactor OutputFactory to introduce various Declaration Visitor classes

### PR DESCRIPTION
**Pull request by sundararajana**
Separate visitors for each activity (duplicate removal, include symbol filtering, name mangling and output generation)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (no review required)

### Issue
 * [CODETOOLS-7903344](https://bugs.openjdk.org/browse/CODETOOLS-7903344): refactor OutputFactory to introduce various Declaration Visitor classes

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jextract pull/81/head:pull/81` \
`$ git checkout pull/81`

Update a local copy of the PR: \
`$ git checkout pull/81` \
`$ git pull https://git.openjdk.org/jextract pull/81/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 81`

View PR using the GUI difftool: \
`$ git pr show -t 81`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jextract/pull/81.diff">https://git.openjdk.org/jextract/pull/81.diff</a>

</details>

### CODE REVIEW

File: samples/libffmpeg/compile.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 jextract -t libffmpeg \
   -I /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include \
-  -I /usr/local/Cellar/ffmpeg@4/4.4.2_3/include \
+  -I /usr/local/Cellar/ffmpeg@4/4.4.2_4/include \
```
Comment (mcimadamore) [inline samples/libffmpeg/compile.sh:3]: Are these changes deliberate?

File: src/main/java/org/openjdk/jextract/impl/CodeGenerator.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+public final class CodeGenerator {
+    private CodeGenerator() {}
+
+    public static JavaFileObject[] generate(Declaration.Scoped decl, String headerName,
+                    String targetPkg, IncludeHelper includeHelper,
+                    List<String> libNames) {
+        var nameMangler = new NameMangler(headerName);
+        var transformedDecl = Stream.of(decl).
+            map(new IncludeFilter(includeHelper)::transform).
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/CodeGenerator.java:40]: Nice pipeline!

File: src/main/java/org/openjdk/jextract/impl/IncludeFilter.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+final class IncludeFilter implements TreeTransformer, Declaration.Visitor<Void, Declaration> {
+    private List<Declaration> decls = new ArrayList<>();
+    private final IncludeHelper includeHelper;
+
+    IncludeFilter(IncludeHelper includeHelper) {
+        this.includeHelper = includeHelper;
+    }
+
+    @Override
+    public Declaration.Scoped transform(Declaration.Scoped header) {
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/IncludeFilter.java:45]: Good to see this stuff factored away

File: src/main/java/org/openjdk/jextract/impl/NameMangler.java [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+/*
+ * This visitor handles java safe names for identifiers, type names. This visitor
+ * stores java safe names in maps. Subsequent code generation steps can check for
+ * java safe names via lookup methods.
+ *
+ * NOTE: Unlike other transforming tree visitors, this one is just identify
+ * transformer. This visitor collects name mappings as it visits tree nodes.
+ * Subsequent code generation steps can check the collected names using getters.
+ */
+final class NameMangler implements TreeTransformer, Declaration.Visitor<Void, Declaration> {
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/NameMangler.java:47]: As discussed offline, there is still coupling between this class and OutputFactory. We have tried several approaches none of which "clicked", so we ended up collecting state in the mangler.

A cleaner solution would be to decorate declarations with "name" attributes, but adding an attribute creates a new declaration, which means there will be types in the tree which would need fixing up. Which is also tricky.

File: src/main/java/org/openjdk/jextract/impl/NameMangler.java [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+import java.util.Optional;
+import java.util.Set;
+import javax.lang.model.SourceVersion;
+
+/*
+ * This visitor handles java safe names for identifiers, type names. This visitor
+ * stores java safe names in maps. Subsequent code generation steps can check for
+ * java safe names via lookup methods.
+ *
+ * NOTE: Unlike other transforming tree visitors, this one is just identify
```
Comment (mcimadamore) [inline src/main/java/org/openjdk/jextract/impl/NameMangler.java:43]: Can't we drop the TreeTransformer implementation?

**Review by mcimadamore — COMMENTED** [review]
The patch improves the code, by separating orthogonal concerns into different visitor passes. It would of course be great if we could go fully there with name mangling as well, but regardless, it looks like a solid step forward (perhaps name mangling can be tackled separately).

File: samples/libffmpeg/compile.sh (hunk 1 above)
Comment (sundararajana) [reply samples/libffmpeg/compile.sh:3]: Yes. Recent home-brew update brought this specific version on my machine. I need this for testing samples.

File: src/main/java/org/openjdk/jextract/impl/NameMangler.java (hunk 5 above)
Comment (sundararajana) [reply src/main/java/org/openjdk/jextract/impl/NameMangler.java]: Removed TreeTransformed implementation for NameMangler.

**Review by mcimadamore — APPROVED** [review]
Looks good
